supplements/tashas-cauldron-of-everything/optional-class-features.xml

```diff
@@ -204,5 +582,1365 @@
+			<extend>ID_PHB_SPELL_REVIVIFY</extend>
+			<extend>ID_WOTC_TCOE_SPELL_SUMMON_FEY</extend>
+			<extend>ID_PHB_SPELL_DOMINATE_BEAST</extend>
+			<extend>ID_WOTC_TCOE_SPELL_SUMMON_ELEMENTAL</extend>
+			<extend>ID_PHB_SPELL_GREATER_RESTORATION</extend>
+		</spellcasting>
+	</element>
+
+	<element name="Ranger, LV02: Additional Ranger Spells" type="Item" source="Tasha’s Cauldron of Everything" id="ID_WOTC_TCOE_ITEM_OCF_RANGER_ADDITIONAL_RANGER_SPELLS">
+		<compendium display="false" />
+		<description>
+			<p><i>You can equip this item to “enable” it. It remains hidden from the inventory on your character sheet.</i></p>
+			<div class="reference">
+				<div element="ID_WOTC_TCOE_CLASS_FEATURE_RANGER_ADDITIONAL_RANGER_SPELLS" />
+			</div>
+		</description>
+		<setters>
+			<set name="category">Optional Class Features</set>
+			<set name="keywords">second level, 2nd level, spellcasting</set>
+			<set name="slot">misc</set>
+			<set name="inventory-hidden">true</set>
+		</setters>
+		<rules>
+			<grant type="Class Feature" id="ID_WOTC_TCOE_CLASS_FEATURE_RANGER_ADDITIONAL_RANGER_SPELLS" requirements="[level:ranger:2]"/>
+		</rules>
+	</element>
+
+	<!-- Spellcasting Focus -->
+	<element name="Spellcasting Focus (Ranger)" type="Class Feature" source="Tasha’s Cauldron of Everything" id="ID_WOTC_TCOE_CLASS_FEATURE_RANGER_SPELLCASTING_FOCUS">
+		<description>
+			<p><i>2nd-level ranger feature</i></p>
+			<p>You can use a druidic focus as a spellcasting focus for your ranger spells. A druidic focus might be a sprig of mistletoe or holly, a wand or rod made of yew or another special wood, a staff drawn whole from a living tree, or an object incorporating feathers, fur, bones, and teeth from sacred animals.</p>
+		</description>
+		<sheet alt="Spellcasting Focus">
+			<description>You can use a druidic focus as a spellcasting focus for your ranger spells.</description>
+		</sheet>
+	</element>
+
+	<element name="Ranger, LV02: Spellcasting Focus (Ranger)" type="Item" source="Tasha’s Cauldron of Everything" id="ID_WOTC_TCOE_ITEM_OCF_RANGER_SPELLCASTING_FOCUS">
+		<compendium display="false" />
+		<description>
+			<p><i>You can equip this item to “enable” it. It remains hidden from the inventory on your character sheet.</i></p>
+			<div class="reference">
+				<div element="ID_WOTC_TCOE_CLASS_FEATURE_RANGER_SPELLCASTING_FOCUS" />
+			</div>
+		</description>
+		<setters>
+			<set name="category">Optional Class Features</set>
+			<set name="keywords">second level, 2nd level, spellcasting</set>
+			<set name="slot">misc</set>
+			<set name="inventory-hidden">true</set>
+		</setters>
+		<rules>
+			<grant type="Class Feature" id="ID_WOTC_TCOE_CLASS_FEATURE_RANGER_SPELLCASTING_FOCUS" requirements="[level:ranger:2]"/>
+		</rules>
+	</element>
+
+	<!-- Primal Awareness -->
+	<element name="Primal Awareness" type="Class Feature" source="Tasha’s Cauldron of Everything" id="ID_WOTC_TCOE_CLASS_FEATURE_RANGER_PRIMAL_AWARENESS">

```

Comment:
For some reason the spells do not show up correctly in magic section or the sheet until the level after it is granted.  On the level the spell is granted the sheet shows it under "Other Spells" as Spell (Primal Awareness).  When level is removed the spell is kept.

This may be a limitation of the app.